components/brave_shields/browser/brave_shields_util.cc

```diff
@@ -336,50 +340,166 @@ DomainBlockingType GetDomainBlockingType(HostContentSettingsMap* map,
+                                        CONTENT_SETTING_DEFAULT);
+      // general-rule:
+      map->SetContentSettingCustomScope(ContentSettingsPattern::Wildcard(),
+                                        host_pattern,
+                                        ContentSettingsType::BRAVE_COOKIES,
+                                        GetDefaultBlockFromControlType(type));
+      break;
+    default:
+      NOTREACHED() << "Invalid ControlType for cookies";
+  }
 
   RecordShieldsSettingChanged(local_state);
 }
+
+struct CookieRules {
+  bool HasDefault() const {
+    return general_setting == CONTENT_SETTING_DEFAULT ||
+           first_party_setting == CONTENT_SETTING_DEFAULT;
+  }
+
+  static CookieRules Get(HostContentSettingsMap* map,
+                         const GURL& url,
+                         ContentSettingsType content_type) {
+    content_settings::SettingInfo general_info;
+    const base::Value& general_value = map->GetWebsiteSetting(
+        GURL::EmptyGURL(), url, content_type, &general_info);
+
+    content_settings::SettingInfo first_party_info;
+    const base::Value& first_party_value =
+        map->GetWebsiteSetting(url, url, content_type, &first_party_info);
+
+    const ContentSettingsPattern& wildcard = ContentSettingsPattern::Wildcard();
+    if (general_info.primary_pattern == wildcard &&
+        general_info.secondary_pattern == wildcard &&
+        first_party_info.primary_pattern == wildcard &&
+        first_party_info.secondary_pattern == wildcard) {
+      return {CONTENT_SETTING_DEFAULT, CONTENT_SETTING_DEFAULT};
+    }
+
+    return {content_settings::ValueToContentSetting(general_value),
+            content_settings::ValueToContentSetting(first_party_value)};
+  }
+
+  static CookieRules GetDefault(
+      content_settings::CookieSettings* cookie_settings) {
+    const ContentSetting default_cookies_setting =
+        cookie_settings->GetDefaultCookieSetting(nullptr);
+    const bool default_should_block_3p_cookies =
+        cookie_settings->ShouldBlockThirdPartyCookies();
+    if (default_cookies_setting == CONTENT_SETTING_BLOCK) {
+      // All cookies are blocked.
+      return {CONTENT_SETTING_BLOCK, CONTENT_SETTING_BLOCK};
+    } else if (default_should_block_3p_cookies) {
+      // First-party cookies are allowed.
+      return {CONTENT_SETTING_BLOCK, CONTENT_SETTING_ALLOW};
+    }
+    // All cookies are allowed.
+    return {CONTENT_SETTING_ALLOW, CONTENT_SETTING_ALLOW};
+  }
+
+  void Merge(const CookieRules& other) {
+    if (general_setting == CONTENT_SETTING_DEFAULT)
+      general_setting = other.general_setting;
+    if (first_party_setting == CONTENT_SETTING_DEFAULT)
+      first_party_setting = other.first_party_setting;
+  }
+
+  ContentSetting general_setting = CONTENT_SETTING_DEFAULT;
+  ContentSetting first_party_setting = CONTENT_SETTING_DEFAULT;
+};
+
 // TODO(bridiver) - convert cookie settings to ContentSettingsType::COOKIES
```

Comment:
I don't think this comment is relevant anymore since we can't use the upstream 3p setting to eliminate having two rules for block 3p